### DISCUSSION TITLE
Switch from C++14 to C++17?

### DISCUSSION THREAD

**Discussion by mloskot** [nanodbc/nanodbc#321] [Polls]
### Now: C++14

Since release v2.12 (in ) nanodbc requires C++14.

The CMake configuration exposes cache variable `CMAKE_CXX_STANDARD` variable with default value `14` (see #303).

Recently, nanodbc received (optional) enhancements that require C++17 and more such may arrive soonish (e.g. support for `std::variant` or `std::any`). There is already an amount of conditionally-compiled code that we maintain (e.g. #148, #283).

The library also uses some C++ features that have been deprecated in C++17, especially the `<codecvt>` (see #74), so those may need to be taken care of soonish too.

### Next: C++17 

If you use or if you plan to use nanodbc, please help us to find out which version of the C++ standard the nanodbc library should support. Please, participate in this poll and discussion thread. Thanks!

/cc @lexicalunit

**Comment by mloskot** [comment]
### C++17 switch rationale

List of reasons why we may want to switch to C++17:

- https://github.com/nanodbc/nanodbc/issues/74
- https://github.com/nanodbc/nanodbc/pull/148
- https://github.com/nanodbc/nanodbc/issues/324
- https://github.com/nanodbc/nanodbc/issues/325
- https://github.com/nanodbc/nanodbc/issues/326
- https://github.com/nanodbc/nanodbc/pull/283
- https://github.com/nanodbc/nanodbc/pull/299
- https://github.com/nanodbc/nanodbc/pull/345
- https://github.com/nanodbc/nanodbc/pull/353#issuecomment-1293130619

**Reply by marlowa** [reply]
C++17 seems like a sensible baseline to me. I attended the C++ ACCU2026 conference this year and a poll showed that many people in corporate environments are stuck on it. That includes me. My work env is rhel8 and the ancient g++ 8.5 there supports C++17 but does not support beyond that. I know that later versions of g++ are available on rhel8 but there is a lot of corporate inertia.

**Reply by lexicalunit** [reply]
Anyone stuck on earlier compiler versions can always use an earlier version of nanodbc as well. Would love to go to C++17 🤘
